plugin-azure-server/src/main/kotlin/jetbrains/buildServer/clouds/azure/arm/KubePodNameGenerator.kt

```diff
@@ -0,0 +1,134 @@
+                           @NotNull  executorServices: ExecutorServices) {
+
+    private val myCounters: ConcurrentMap<String, AtomicInteger> = ConcurrentHashMap()
+    private val myIndexTouchedMaps: ConcurrentMap<String, AtomicBoolean> = ConcurrentHashMap()
+    private var myIndexStorage: File = File(ServerPaths.PLUGIN_DATA_DIRECTORY_NAME, "azureKubeIndex")
+    private val myIsAvailable: AtomicBoolean = AtomicBoolean(false)
+    private val myLock: ReadWriteLock = ReentrantReadWriteLock()
+
+    init{
+        if (!myIndexStorage.exists()) {
+            myIndexStorage.mkdirs()
+        }
+        if (!myIndexStorage.isDirectory) {
+            throw CloudException("Unable to create a directory for azure arm plugin to monitor VM indexes")
+        }
+        myIsAvailable.set(true)
+
+        loadIndexes(myIndexStorage)
+
+        eventDispatcher.addListener(object : BuildServerAdapter() {
+            override fun serverShutdown() {
+                myIsAvailable.set(false)
+                storeIndexes()
+            }
+        })
+        executorServices.normalExecutorService.scheduleWithFixedDelay(this::storeIndexes, 60, 60, TimeUnit.SECONDS)
+    }
+
+    @Synchronized
+    private fun loadIndexes(@NotNull indexStorage: File) {
+        val indexes = indexStorage.listFiles() ?: return
+        for (indexFile in indexes) {
+            if (!indexFile.name.endsWith(".idx")) continue
+            val indexName = indexFile.name.substring(0, indexFile.name.length - 4)
+            try {
+                val indexValueFromFile = FileUtil.readText(indexFile).toInt()
+                if (indexValueFromFile > 0) {
+                    myCounters.putIfAbsent(indexName, AtomicInteger(0))
+                    myCounters[indexName]?.set(indexValueFromFile)
+                }
+            } catch (e: IOException) {
+                continue
+            }
+        }
+    }
+
+    private fun storeIndexes() {
+        // wait for generation operations to finish:
+        try {
+            if (!myLock.writeLock().tryLock(100, TimeUnit.MILLISECONDS)) {
+                Loggers.AGENT.warn("Waited more than 100ms to store Kube indexes")
+            }
+            for ((key, value) in myIndexTouchedMaps) {
+                if (value.compareAndSet(true, false)) {
+                    val counter = myCounters[key]
+                    try {
+                        val indexFile = File(myIndexStorage, "$key.idx")
+                        FileUtil.writeFile(indexFile,counter.toString(), Charset.defaultCharset().name())
+                    } catch (ignored: IOException) {
+                        continue
+                    }
+                }
+            }
+        } catch (e: InterruptedException) {
+            e.printStackTrace()
+        } finally {
+            myLock.writeLock().unlock()
+        }
+    }
+
+    fun generateNewVmName(image: AzureCloudImage): String {
+        try {
+            myLock.readLock().lock()
```

Comment:
I think `lock()` blocks forever it might be better to just set a timeout of ten minutes rather than risk holding a thread forever.

Reply:
Again,  locks can through an InterruptedException, they need to be inside the try block.